nemo/collections/nlp/models/machine_translation/transformer_mt_model.py

```diff
@@ -0,0 +1,332 @@
+            attn_layer_dropout=cfg.machine_translation.attn_layer_dropout,
+        )
+        self.log_softmax = TokenClassifier(
+            hidden_size=cfg.machine_translation.hidden_size, num_classes=tgt_vocab_size, log_softmax=True,
+        )
+        self.beam_search = BeamSearchSequenceGenerator(
+            embedding=self.tgt_embedding_layer,
+            decoder=self.decoder,
+            log_softmax=self.log_softmax,
+            max_sequence_length=cfg.machine_translation.max_seq_length,
+            beam_size=cfg.machine_translation.beam_size,
+            bos=self.tgt_tokenizer.bos_id,
+            pad=self.tgt_tokenizer.pad_id,
+            eos=self.tgt_tokenizer.eos_id,
+            len_pen=cfg.machine_translation.len_pen,
+            max_delta_length=cfg.machine_translation.get("max_generation_delta", 50),
+        )
+
+        std_init_range = 1 / cfg.machine_translation.hidden_size ** 0.5
+        self.apply(lambda module: transformer_weights_init(module, std_init_range))
+
+        # tie weights of embedding and softmax matrices
+        self.log_softmax.mlp.layer0.weight = self.tgt_embedding_layer.token_embedding.weight
+        self.emb_scale = cfg.machine_translation.hidden_size ** 0.5
+        self.loss_fn = SmoothedCrossEntropyLoss(
+            pad_id=self.tgt_tokenizer.pad_id, label_smoothing=cfg.machine_translation.label_smoothing
+        )
+
+        # Optimizer setup needs to happen after all model weights are ready
+        self.setup_optimization(cfg.optim)
+
+        # These attributes are added to bypass Illegal memory access error in PT1.6
+        # https://github.com/pytorch/pytorch/issues/21819
+
+    @typecheck()
+    def forward(self, src, src_mask, tgt, tgt_mask):
+        """
+        torch.nn.Module.forward method.
+        Args:
+            src: source ids
+            src_mask: src mask (mask padding)
+            tgt: target ids
+            tgt_mask: target mask
+
+        Returns:
+
+        """
+        src_embeddings = self.src_embedding_layer(input_ids=src)
+        # src_embeddings *= src_embeddings.new_tensor(self.emb_scale)
+        src_hiddens = self.encoder(src_embeddings, src_mask)
+        tgt_embeddings = self.tgt_embedding_layer(input_ids=tgt)
+        # tgt_embeddings *= tgt_embeddings.new_tensor(self.emb_scale)
+        tgt_hiddens = self.decoder(tgt_embeddings, tgt_mask, src_hiddens, src_mask)
+        log_probs = self.log_softmax(hidden_states=tgt_hiddens)
+        beam_results = None
+        if not self.training:
+            beam_results = self.beam_search(encoder_hidden_states=src_hiddens, encoder_input_mask=src_mask)
+        return log_probs, beam_results
+
+    def training_step(self, batch, batch_idx):
+        """
+        Lightning calls this inside the training loop with the data from the training dataloader
+        passed in as `batch`.
+        """
+        # forward pass
+        for i in range(len(batch)):
+            if batch[i].ndim == 3:
```

Comment:
If batch is a valid tensor, then every `batch[i]` should have the same shape, so why do you check for this? Also, why not just squeeze the batch tensor directly?

Reply:
I thought that there could be another dataset that does not batch data. If such a dataset is used, `batch[i].ndim == 2` and no squeezing is needed.

`batch` is not a tensor but a tuple of tensors.

Reply:
I see, makes sense!